packages/app/app/components/Settings/index.js

```diff
@@ -77,10 +81,87 @@ class Settings extends React.Component {
       >
         {this.renderLastFmLoginButtons()}
         {this.renderLastFmOptionRadio()}
+        {this.renderLastFmImportFavButton()}
       </SocialIntegration>
     );
   }
+  UpdateImportMsg(msg) {
+    this.setState({
+      lastFmFavImportMessage: msg
+    });
+  }
+  renderLastFmImportFavButton(){
+    let {
+      lastFmName,
+      lastFmSessionKey,
+      lastFmFavImportStatus
+    } = this.props.scrobbling;
 
+    let {
+      FmFavImport
+    } = this.props.actions;
+    return ( lastFmName && lastFmSessionKey && (
+      <div className={styles.settings_social_item}>
+        <span>
+          {this.props.t('fmfav-msg')}
+        </span>
+        <Spacer />
+        <span>
+          <strong>{this.state.lastFmFavImportMessage}</strong>
+        </span>
+        <Spacer />
+        {(
+          <Button disabled={!lastFmFavImportStatus} loading={!lastFmFavImportStatus} onClick={() => {
```

Comment:
This is way too complex for a handler. You should move all this logic to an action, but refactored to be simpler.

You can follow the pattern we have in many other places, with actions ending in `_START`, `_SUCCESS`, and `_ERROR` to update the redux state with data accordingly. You can avoid most of this branching logic this way. And then you can join these simple actions into a sequence using a thunk.

Let me know if I should make this clearer or if you need an example.